hardware/src/libs/ESP8266/ESP8266.cpp

```diff
@@ -0,0 +1,721 @@
+bool ESP8266::releaseTCP(void)
+{
+    return eATCIPCLOSESingle();
+}
+
+bool ESP8266::registerUDP(String addr, uint32_t port)
+{
+    return sATCIPSTARTSingle("UDP", addr, port);
+}
+
+bool ESP8266::unregisterUDP(void)
+{
+    return eATCIPCLOSESingle();
+}
+
+bool ESP8266::createTCP(uint8_t mux_id, String addr, uint32_t port)
+{
+    return sATCIPSTARTMultiple(mux_id, "TCP", addr, port);
+}
+
+bool ESP8266::releaseTCP(uint8_t mux_id)
+{
+    return sATCIPCLOSEMulitple(mux_id);
+}
+
+bool ESP8266::registerUDP(uint8_t mux_id, String addr, uint32_t port)
+{
+    return sATCIPSTARTMultiple(mux_id, "UDP", addr, port);
+}
+
+bool ESP8266::unregisterUDP(uint8_t mux_id)
+{
+    return sATCIPCLOSEMulitple(mux_id);
+}
+
+bool ESP8266::setTCPServerTimeout(uint32_t timeout)
+{
+    return sATCIPSTO(timeout);
+}
+
+bool ESP8266::startTCPServer(uint32_t port)
+{
+    if (sATCIPSERVER(1, port)) {
+        return true;
+    }
+    return false;
+}
+
+bool ESP8266::stopTCPServer(void)
+{
+    sATCIPSERVER(0);
+    restart();
+    return false;
+}
+
+bool ESP8266::startServer(uint32_t port)
+{
+    return startTCPServer(port);
+}
+
+bool ESP8266::stopServer(void)
+{
+    return stopTCPServer();
+}
+
+bool ESP8266::send(const uint8_t *buffer, uint32_t len)
+{
+    return sATCIPSENDSingle(buffer, len);
+}
+
+bool ESP8266::send(uint8_t mux_id, const uint8_t *buffer, uint32_t len)
+{
+    return sATCIPSENDMultiple(mux_id, buffer, len);
+}
+
+uint32_t ESP8266::recv(uint8_t *buffer, uint32_t buffer_size, uint32_t timeout)
+{
+    return recvPkg(buffer, buffer_size, NULL, timeout, NULL);
+}
+
+uint32_t ESP8266::recv(uint8_t mux_id, uint8_t *buffer, uint32_t buffer_size, uint32_t timeout)
+{
+    uint8_t id;
+    uint32_t ret;
+    ret = recvPkg(buffer, buffer_size, NULL, timeout, &id);
+    if (ret > 0 && id == mux_id) {
+        return ret;
+    }
+    return 0;
+}
+
+uint32_t ESP8266::recv(uint8_t *coming_mux_id, uint8_t *buffer, uint32_t buffer_size, uint32_t timeout)
+{
+    return recvPkg(buffer, buffer_size, NULL, timeout, coming_mux_id);
+}
+
+/*----------------------------------------------------------------------------*/
+/* +IPD,<id>,<len>:<data> */
+/* +IPD,<len>:<data> */
+
+uint32_t ESP8266::recvPkg(uint8_t *buffer, uint32_t buffer_size, uint32_t *data_len, uint32_t timeout, uint8_t *coming_mux_id)
+{
+    String data;
+    char a;
+    int32_t index_PIPDcomma = -1;
+    int32_t index_colon = -1; /* : */
+    int32_t index_comma = -1; /* , */
+    int32_t len = -1;
+    int8_t id = -1;
+    bool has_data = false;
+    uint32_t ret;
+    unsigned long start;
+    uint32_t i;
+
+    if (buffer == NULL) {
+        return 0;
+    }
+
+    start = millis();
+    while (millis() - start < timeout) {
+        if(m_puart->available() > 0) {
+            a = m_puart->read();
+            data += a;
+        }
+
+        index_PIPDcomma = data.indexOf("+IPD,");
+        if (index_PIPDcomma != -1) {
+            index_colon = data.indexOf(':', index_PIPDcomma + 5);
```

Comment:
Not actually my code. Just realised I haven't linked the library anywhere though